src/orderings.jl

```diff
@@ -82,4 +93,41 @@ function lt(o::WreathOrder, p::T, q::T) where T<:AbstractWord{<:Integer}
     first_pp = findfirst(isequal(max_p), pp)
     first_qq = findfirst(isequal(max_p), qq)
     return @views lt(o, pp[1:first_pp - 1], qq[1:first_qq - 1])
-end
\ No newline at end of file
+end
+
+
+"""
+    struct RecursivePathOrder{T} <: WordOrdering
+
+Structure representing Recursive Path Ordering (determined by the Lexicographic
+ordering of the Alphabet) of the words over given Alphabet. This Lexicographinc
+ordering of an Alphabet is implicitly specified inside Alphabet struct.
+"""
+struct RecursivePathOrder{T} <: WordOrdering
+    A::Alphabet{T}
+end
+
+Base.hash(o::RecursivePathOrder, h::UInt) = hash(o.A, hash(h, hash(RecursivePathOrder)))
+
+"""
+    lt(o::RecursivePathOrder, p::T, q::T) where T<:AbstractWord{<:Integer}
+
+Return whether the first word is less then the other one in a given
+RecursivePathOrder ordering.
+"""
+function lt(o::RecursivePathOrder, p::T, q::T) where T<:AbstractWord{<:Integer}
+    isone(p) && return !isone(q)
+    isone(q) && return false
+
+    # i.e. p, q ≂̸ ε
+    length(p) == 1 && length(q) == 1 && return (first(p) < first(q))
+
+    # i.e. we are not comparing single letter words
+    if first(p) == first(q)
+        @views lt(o, p[2:end], q[2:end]) && return true
+    elseif first(p) < first(q)
+        @views lt(o, p[2:end], q[1:end]) && return true
+    end
+    @views p == q[2:end] && return true
+    @views return lt(o, p[1:end], q[2:end])
```

Comment:
this part is a little bit unclear to me; the direct translation of rpo from the paper (as I understand it) would be 
```julia
function lt(o::RecursivePathOrder, p::T, q::T) where T<:AbstractWord{<:Integer}
    [...] # deal with exceptional cases here
    # p <_rpo q iff one of the following holds 
    @views begin
        lt(o, p, q[2:end]) && return true
        first(p) < first(q) && lt(o, p[2:end], q) && return true
        first(p) == first(q) && lt(o, p[2:end], q[2:end]) && return true
    end
    return false
end
```
is your version equivalent to this?

Reply:
I believe you are not asking about my `if ... elseif ...` (this works similar to your code, and I agree that your code looks nicer) and rather about why the other condition is outside it and why it has equality first. As to equality: note that the condition (1) in the paper is given as `<=` (or to be precise `>=`).  So in your code we would need to add the `==` condition (and it is indeed needed, as otherwise the ordering will not be compatible with concatenation - I will add a test visualising this as the current one is not enough). Except for that your code should give the same result (with one technical change: as far as I know `@views` is used only when you slice an array - i.e. works for `a[1:end]` but not `a` itself. So we either need to use `a[1:end]` or change `lt(o::RecursivePathOrder, p::T, q::T) where ...` to `lt(o::RecursivePathOrder, p::T, q::S) where ...` - whichever you prefer).

One remark: I did try to understand this ordering and I believe I do not understand it. I do not know how to order the set of words just by looking at it. I know how to check that it is not implemented properly: namely check if it is compatible with concatenation via making sure that subwords of a word are smaller than the word itself.

Reply:
Oh, that's a good catch with `≥_rpo` in the paper! So the correct version should read:
```julia
    @views begin
        (p == q[2:end] || lt(o, p, q[2:end])) && return true
        first(p) < first(q) && lt(o, p[2:end], q) && return true
        first(p) == first(q) && lt(o, p[2:end], q[2:end]) && return true
    end
```
??

as for the signature of `lt` I'd prefer
```
lt(o::RecursivePathOrder, p::AbstractWord, q::AbstractWord)
```
except for enforcing equality of types, `T` does nothing in the body

Reply:
This is correct or we can just have explicit four conditions like I updated in now:
```julia
    @views begin
        p[1:end] == q[2:end] && return true
        lt(o, p[1:end], q[2:end]) && return true
        first(p) < first(q) && lt(o, p[2:end], q[1:end]) && return true
        first(p) == first(q) && lt(o, p[2:end], q[2:end]) && return true
    end
```
and I think four conditions are even clearer for the reading as they somehow show better (to the reader) that the subwords are `<_rpo` the the whole word.

I actually had the type enforcing because `lt` works only when `AbstractWord{T}` has `T <: Integer` but I looked again and it seems that there is no way for `T` to be something else. So I will update it the way you suggested.